File: scripts/download_images.py
import csv
import json
import os
import urllib.request
import time
from pathlib import Path

# Read CSV file
csv_path = r'c:\Users\migut\Desktop\workout_data (1).csv'
print(f"Reading CSV from: {csv_path}")

exercises = set()
try:
    with open(csv_path, 'r', encoding='utf-8') as f:
        reader = csv.DictReader(f)
        for row in reader:
            if 'exercise_title' in row and row['exercise_title']:
                exercises.add(row['exercise_title'])
except Exception as e:
    print(f"Error reading CSV: {e}")
    exit(1)

exercise_list = sorted(list(exercises))
print(f"\nFound {len(exercise_list)} unique exercises:")
for i, exercise in enumerate(exercise_list, 1):
    print(f"  {i}. {exercise}")

# Create images directory
images_dir = Path(r'c:\Users\migut\migutse Gym\client\public\exercise-images')
images_dir.mkdir(parents=True, exist_ok=True)
print(f"\nCreated directory: {images_dir}")

# Download images
UNSPLASH_API_KEY = 'BxKxgKxSa2JKGjlXc6Mg_MdNH7FfZZXAYs9nfhTGNYo'
image_mapping = {}
download_count = 0
fail_count = 0

print(f"\nDownloading images for {len(exercise_list)} exercises...\n")

for exercise in exercise_list:
    try:
        # Fetch from Unsplash
        url = f"https://api.unsplash.com/search/photos?query={urllib.parse.quote(exercise)}&per_page=1&orientation=landscape&client_id={UNSPLASH_API_KEY}"
        
        with urllib.request.urlopen(url, timeout=10) as response:
            data = json.loads(response.read().decode('utf-8'))
            
            if data.get('results') and len(data['results']) > 0:
                image_url = data['results'][0]['urls']['regular']
                image_name = f"{exercise.lower().replace(' ', '-')}.jpg"
                image_path = images_dir / image_name
                
                try:
                    print(f"Downloading: {exercise}...", end=' ')
                    urllib.request.urlretrieve(image_url, str(image_path))
                    image_mapping[exercise] = f"/exercise-images/{image_name}"
                    download_count += 1
                    print("✓")
                except Exception as e:
                    fail_count += 1
                    print(f"✗ (failed to save)")
            else:
                fail_count += 1
                print(f"✗ {exercise} - no image found")
            
            # Rate limiting
            time.sleep(0.2)
    except Exception as e:
        fail_count += 1
        print(f"✗ {exercise} - {str(e)[:50]}")

# Save mapping file
mapping_path = Path(r'c:\Users\migut\migutse Gym\client\public\exercise-images.json')
with open(mapping_path, 'w', encoding='utf-8') as f:
    json.dump(image_mapping, f, indent=2)

print(f"\n✓ Saved mapping to: {mapping_path}")
print(f"\nSummary: Downloaded {download_count}, Failed {fail_count}")

File: download-images.js
#!/usr/bin/env node

const fs = require('fs');
const path = require('path');
const https = require('https');

// Common exercises based on typical Hevy exports
const exercises = [
  'Barbell Bench Press',
  'Dumbbell Bench Press',
  'Barbell Squat',
  'Leg Press',
  'Barbell Deadlift',
  'Romanian Deadlift',
  'Leg Curl',
  'Leg Extension',
  'Barbell Row',
  'Dumbbell Row',
  'Lat Pulldown',
  'Barbell Overhead Press',
  'Dumbbell Shoulder Press',
  'Cable Machine Chest Fly',
  'Dumbbell Lateral Raise',
  'Barbell Curl',
  'Dumbbell Curl',
  'Barbell Tricep Extension',
  'Cable Tricep Pushdown',
  'Tricep Rope Pushdown',
  'Leg Raises',
  'Ab Wheel',
  'Cable Wood Chop',
  'Pull-ups',
  'Chin-ups',
  'Machine Leg Press',
  'Smith Machine Bench Press',
  'Incline Dumbbell Press',
  'Dumbbell Flyes',
  'Machine Leg Curl',
  'Hack Squat',
  'Walking Lunges',
  'Bulgarian Split Squat',
  'Leg Press',
  'Cable Fly',
  'Chest Press Machine',
  'Seated Row',
  'Back Extension',
  'Cable Pull Through',
  'Glute Kickback',
  'Hip Thrust',
  'Plate Loaded Leg Press',
  'Smith Machine Squat',
  'Machine Shoulder Press',
  'Lateral Raise Machine',
  'Tricep Dip Machine',
  'Leg Curl Machine'
];

const imagesDir = path.join(__dirname, '..', 'client', 'public', 'exercise-images');
if (!fs.existsSync(imagesDir)) {
  fs.mkdirSync(imagesDir, { recursive: true });
}

const UNSPLASH_API_KEY = 'BxKxgKxSa2JKGjlXc6Mg_MdNH7FfZZXAYs9nfhTGNYo';
let imageMapping = {};
let completed = 0;

function downloadImage(url, filepath) {
  return new Promise((resolve) => {
    https.get(url, (response) => {
      const file = fs.createWriteStream(filepath);
      response.pipe(file);
      file.on('finish', () => {
        file.close();
        resolve(true);
      });
      file.on('error', () => resolve(false));
    }).on('error', () => resolve(false));
  });
}

async function fetchImage(exercise) {
  return new Promise((resolve) => {
    const url = `https://api.unsplash.com/search/photos?query=${encodeURIComponent(exercise)}&per_page=1&orientation=landscape&client_id=${UNSPLASH_API_KEY}`;
    
    https.get(url, (response) => {
      let data = '';
      response.on('data', chunk => data += chunk);
      response.on('end', async () => {
        try {
          const json = JSON.parse(data);
          if (json.results && json.results.length > 0) {
            resolve(json.results[0].urls.regular);
          } else {
            resolve(null);
          }
        } catch (e) {
          resolve(null);
        }
      });
    }).on('error', () => resolve(null));
  });
}

async function downloadAllImages() {
  console.log(`\nDownloading ${exercises.length} exercise images...\n`);

  for (const exercise of exercises) {
    const imageUrl = await fetchImage(exercise);
    
    if (imageUrl) {
      const imageName = `${exercise.toLowerCase().replace(/\s+/g, '-')}.jpg`;
      const imagePath = path.join(imagesDir, imageName);
      
      const success = await downloadImage(imageUrl, imagePath);
      if (success) {
        imageMapping[exercise] = `/exercise-images/${imageName}`;
        console.log(`✓ ${exercise}`);
      } else {
        console.log(`✗ ${exercise} (failed to save)`);
      }
    } else {
      console.log(`✗ ${exercise} (no image found)`);
    }
    
    completed++;
    // Rate limiting
    await new Promise(resolve => setTimeout(resolve, 300));
  }

  // Save mapping
  const mappingPath = path.join(__dirname, '..', 'client', 'public', 'exercise-images.json');
  fs.writeFileSync(mappingPath, JSON.stringify(imageMapping, null, 2));
  
  console.log(`\n✓ Saved ${Object.keys(imageMapping).length} images to ${mappingPath}`);
}

downloadAllImages();
